### PR TITLE
GH-1497 added 'file' URI validation

### PR DESCRIPTION
**Pull request by olegz**
- added 'file' URI validation to AppRegistry
- added initial version of AppRegistryTests
- fixed AppRegistryControllerTests which were failing due to this change

### CODE REVIEW

File: spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java [hunk 1]
```diff
@@ -235,4 +238,12 @@ else if (!StringUtils.hasText(uri.getSchemeSpecificPart())) {
 		}
 		return uri;
 	}
+
+	private void validateApplicationUri(URI uri){
+		if (uri.getScheme().toLowerCase().equals("file")){
+			File file = new File(uri);
+			Assert.isTrue(file.exists() && !file.isDirectory() && file.getName().toLowerCase().endsWith(".jar"),
```
Comment (ericbottard) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java]: File *can* be a directory (we support exploded archives)

Comment (ericbottard) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java]: Moreover, file *could* be something else than a boot app (could be a python app ultimately, etc)

File: spring-cloud-dataflow-registry/src/test/java/org/springframework/cloud/dataflow/registry/AppRegistryTests.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		AppRegistration registration = this.appRegistry.save("foo", ApplicationType.processor, uri, null);
+		assertThat(registration != null);
+		assertEquals("foo", registration.getName());
+		assertEquals(ApplicationType.processor, registration.getType());
+		assertEquals(uri, registration.getUri());
+	}
+
+	@Test
+	public void testWithExistingJARFileURI() throws Exception {
+		File jar = new File("foo.jar");
```
Comment (ericbottard) [inline spring-cloud-dataflow-registry/src/test/java/org/springframework/cloud/dataflow/registry/AppRegistryTests.java]: Should reside in temporary ad hoc directory, not working directory. See JUnit rule for that purpose.

Applies to all tests in this class

File: spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java (hunk 1 above)
Comment (olegz) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java]: Hmm, that changes things. Wasn't aware of that. Perhaps I need to revisit the approach all together to at least ensure that that user has a better way of being notified that the final stream can not be deployed due to. . . .

**Comment by olegz** [comment]
@ericbottard Thanks for the pointers.  I am closing this PR for now based on the file support comment. Will submit the new one once i figure out how to make it more user friendly.

**Pull request closed without merging**